pkg/proxy/iptables/proxier.go

```diff
@@ -1099,6 +1126,52 @@ func (proxier *Proxier) syncProxyRules() {
 			activeNATChains[svcXlbChain] = true
 		}
 
+		if proxier.localDetector.IsImplemented() && svcChain != "" {
+			// We have a LocalDetector, so fill in the rules in the SVC chain to
+			// redirect external traffic (or simply masquerade it if we don't
+			// have a separate external chain).
+			var externalTrafficDest utiliptables.Chain
+			if svcXlbChain != "" {
+				externalTrafficDest = svcXlbChain
+			} else {
+				externalTrafficDest = KubeMarkMasqChain
+			}
+
+			// Note that this matches traffic where both src-type and the
+			// LocalDetector consider it non-local. ie, it's not a node IP and
+			// it's not a pod IP.
+			proxier.natRules.Write(
+				"-A", string(svcChain),
+				"-m", "comment", "--comment", svcNameString,
+				"-m", "addrtype",
+				"!", "--src-type", "LOCAL",
+				proxier.localDetector.JumpIfNotLocal(nil, string(externalTrafficDest)),
+			)
+		} else if !proxier.localDetector.IsImplemented() && svcXlbChain != "" {
+			// No LocalDetector, so fill in the rules in the XLB chain to treat
+			// node-local traffic as internal
+			proxier.natRules.Write(
+				"-A", string(svcXlbChain),
+				"-m", "comment", "--comment", fmt.Sprintf(`"masquerade LOCAL traffic for %s LB IP"`, svcNameString),
+				"-m", "addrtype", "--src-type", "LOCAL",
+				"-j", string(KubeMarkMasqChain),
+			)
+			proxier.natRules.Write(
```

Comment:
Am I reading this right?  In the first case (above) you have a rule in svcChain which jumps to xlbChain.  In this else-if you have a rule in xlbChain that jumps to svcChain.

?

That seems like it could end up in a loop ?  Or at least be impossible to diagram...

Reply:
There's no loop because you can only have the SVC->XLB jump if you have a LocalDetector, and you can only have the XLB->SVC jump if you _don't_ have a LocalDetector.